From c7c10933b743a29cc785edd0aa1824c19e53a7df Mon Sep 17 00:00:00 2001
From: Oliver Scott <olivercscott@gmail.com>
Date: Wed, 17 Apr 2024 23:31:13 -0400
Subject: [PATCH] Camera timeout feature

Ported to AXP.OS lineage-22.2 (rhodep fork) from GrapheneOS 2578144de3a7.

Change-Id: I146a6b7549983009df13f4af7e1028c0cc7accc1
---
 core/java/android/provider/Settings.java      |  6 +++
 .../settings/backup/SecureSettings.java       |  1 +
 .../validators/SecureSettingsValidators.java  | 15 ++++++
 .../privacy/AppOpsPrivacyItemMonitor.kt       | 54 +++++++++++++++++--
 4 files changed, 73 insertions(+), 3 deletions(-)

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 2b196face..6cec95b08 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -12911,6 +12911,12 @@ public final class Settings {
          */
         public static final String HIDE_PRIVATESPACE_ENTRY_POINT = "hide_privatespace_entry_point";
 
+        /**
+         * The amount of time in milliseconds before camera is turned off
+         * @hide
+         */
+        public static final String CAMERA_OFF_TIMEOUT = "camera_off_timeout";
+
         /**
          * Whether or not secure windows should be disabled. This only works on debuggable builds.
          *
diff --git a/packages/SettingsProvider/src/android/provider/settings/backup/SecureSettings.java b/packages/SettingsProvider/src/android/provider/settings/backup/SecureSettings.java
index 0944766d4..55122a3e9 100644
--- a/packages/SettingsProvider/src/android/provider/settings/backup/SecureSettings.java
+++ b/packages/SettingsProvider/src/android/provider/settings/backup/SecureSettings.java
@@ -289,5 +289,6 @@ public class SecureSettings {
         Settings.Secure.MANDATORY_BIOMETRICS_REQUIREMENTS_SATISFIED,
         Settings.Secure.ADVANCED_PROTECTION_MODE,
         Settings.Secure.ACCESSIBILITY_KEY_GESTURE_TARGETS,
+        Settings.Secure.CAMERA_OFF_TIMEOUT
     };
 }
diff --git a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
index 0c5d3a426..c691b55cb 100644
--- a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
+++ b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
@@ -452,5 +452,20 @@ public class SecureSettingsValidators {
         VALIDATORS.put(Secure.MANDATORY_BIOMETRICS_REQUIREMENTS_SATISFIED,
                 new InclusiveIntegerRangeValidator(0, 1));
         VALIDATORS.put(Secure.ADVANCED_PROTECTION_MODE, BOOLEAN_VALIDATOR);
+        VALIDATORS.put(Secure.CAMERA_OFF_TIMEOUT,
+                new DiscreteValueValidator(new String[] {
+                        "0",
+                        "15000",
+                        "30000",
+                        "60000",
+                        "120000",
+                        "300000",
+                        "600000",
+                        "1800000",
+                        "3600000",
+                        "7200000",
+                        "14400000",
+                        "28800000"
+                }));
     }
 }
diff --git a/packages/SystemUI/src/com/android/systemui/privacy/AppOpsPrivacyItemMonitor.kt b/packages/SystemUI/src/com/android/systemui/privacy/AppOpsPrivacyItemMonitor.kt
index fedbdec11..0f40aabd7 100644
--- a/packages/SystemUI/src/com/android/systemui/privacy/AppOpsPrivacyItemMonitor.kt
+++ b/packages/SystemUI/src/com/android/systemui/privacy/AppOpsPrivacyItemMonitor.kt
@@ -16,10 +16,16 @@
 
 package com.android.systemui.privacy
 
+import android.app.AlarmManager
+import android.app.AlarmManager.OnAlarmListener
 import android.app.AppOpsManager
 import android.content.Context
 import android.content.pm.UserInfo
+import android.database.ContentObserver
+import android.hardware.SensorPrivacyManager
+import android.os.SystemClock
 import android.os.UserHandle
+import android.provider.Settings
 import com.android.internal.annotations.GuardedBy
 import com.android.internal.annotations.VisibleForTesting
 import com.android.systemui.appops.AppOpItem
@@ -44,6 +50,7 @@ import javax.inject.Inject
  */
 @SysUISingleton
 class AppOpsPrivacyItemMonitor @Inject constructor(
+    private val context: Context,
     private val appOpsController: AppOpsController,
     private val userTracker: UserTracker,
     private val privacyConfig: PrivacyConfig,
@@ -53,14 +60,16 @@ class AppOpsPrivacyItemMonitor @Inject constructor(
 
     @VisibleForTesting
     companion object {
-        val OPS_MIC_CAMERA = intArrayOf(
-                AppOpsManager.OP_CAMERA,
-                AppOpsManager.OP_PHONE_CALL_CAMERA,
+        private val OPS_MIC = intArrayOf(
                 AppOpsManager.OP_RECORD_AUDIO,
                 AppOpsManager.OP_PHONE_CALL_MICROPHONE,
                 AppOpsManager.OP_RECEIVE_AMBIENT_TRIGGER_AUDIO,
                 AppOpsManager.OP_RECEIVE_EXPLICIT_USER_INTERACTION_AUDIO,
                 AppOpsManager.OP_RECEIVE_SANDBOX_TRIGGER_AUDIO)
+        private val OPS_CAMERA = intArrayOf(
+                AppOpsManager.OP_CAMERA,
+                AppOpsManager.OP_PHONE_CALL_CAMERA)
+        val OPS_MIC_CAMERA = OPS_MIC + OPS_CAMERA
         val OPS_LOCATION = intArrayOf(
                 AppOpsManager.OP_COARSE_LOCATION,
                 AppOpsManager.OP_FINE_LOCATION)
@@ -95,6 +104,9 @@ class AppOpsPrivacyItemMonitor @Inject constructor(
                 if (code in OPS_LOCATION && !locationAvailable) {
                     return
                 }
+                if (code in OPS_CAMERA && !active) {
+                    setCameraTimeout()
+                }
                 if (userTracker.userProfiles.any { it.id == UserHandle.getUserId(uid) } ||
                         code in USER_INDEPENDENT_OPS) {
                     logger.logUpdatedItemFromAppOps(code, uid, packageName, active)
@@ -134,10 +146,46 @@ class AppOpsPrivacyItemMonitor @Inject constructor(
         }
     }
 
+    private val cameraTimeoutListener = OnAlarmListener {
+        if (getActivePrivacyItems().none { it.privacyType == PrivacyType.TYPE_CAMERA }) {
+            context.getSystemService(SensorPrivacyManager::class.java)
+                ?.setSensorPrivacy(SensorPrivacyManager.Sensors.CAMERA, true)
+        }
+    }
+
     init {
+        setCameraTimeout()
+        context.contentResolver.registerContentObserver(Settings.Secure.getUriFor(
+                Settings.Secure.CAMERA_OFF_TIMEOUT), false,
+                object : ContentObserver(null) {
+                    override fun onChange(selfChange: Boolean) {
+                        setCameraTimeout()
+                    }
+                })
+        context.getSystemService(SensorPrivacyManager::class.java)
+            ?.addSensorPrivacyListener { sensor, enabled ->
+                if (!enabled) {
+                    when (sensor) {
+                        SensorPrivacyManager.Sensors.CAMERA -> setCameraTimeout()
+                    }
+                }
+            }
         privacyConfig.addCallback(configCallback)
     }
 
+    private fun setCameraTimeout() {
+        val cameraTimeoutMillis: Long = Settings.Secure.getLong(context.getContentResolver(),
+                Settings.Secure.CAMERA_OFF_TIMEOUT, 0)
+        val alarmManager: AlarmManager? = context.getSystemService(AlarmManager::class.java)
+        alarmManager?.cancel(cameraTimeoutListener)
+        if (cameraTimeoutMillis != 0L) {
+            val timeout: Long = SystemClock.elapsedRealtime() + cameraTimeoutMillis
+            alarmManager?.setExactAndAllowWhileIdle(AlarmManager.ELAPSED_REALTIME_WAKEUP, timeout,
+                    AppOpsPrivacyItemMonitor::class.java.simpleName,
+                    Runnable::run, null, cameraTimeoutListener)
+        }
+    }
+
     override fun startListening(callback: PrivacyItemMonitor.Callback) {
         synchronized(lock) {
             this.callback = callback
-- 
2.43.0

